Menendez develop

Menendez2004 opened the pull request:
all challenge are already done, front-end
- challenge1
- challenge2 
 and Js-algorithms are already done

Menendez2004 committed:
Im already finish the challenges
All front end are already did it
Also the jS-Algorithms are already done

Menendez2004 committed:
correction of the last commit, i didn´t add the js algorithms
redeableTime challenge done
country_name challenge done
ownPower challenge done
digitSum challenge done
fibindex challenge done

erickvh commented on JS-Algorithms/challenges.js:
```diff
@@ -14,6 +14,14 @@ Invoking "readableTime(3690)" should return "01:01:30" (HH:MM:SS)
+  let secs = Math.floor(seconds % 3600 % 60).toString().padStart(2, '0');//literally the same as above but obtain the module of the remaining seconds by 60 to get the seconds
+  // console.log(hours)//just for check iff the format is corrct xd
+  // console.log(minutes)
+  // console.log(secs)
```
Don't forget remove unused debug code and comments.

erickvh commented on JS-Algorithms/challenges.js:
```diff
@@ -14,6 +14,14 @@ Invoking "readableTime(3690)" should return "01:01:30" (HH:MM:SS)
 
 const readableTime = (seconds) => {
   // YOUR CODE HERE...
```
This line it is supposed to be removed

erickvh commented on JS-Algorithms/challenges.js:
```diff
@@ -14,6 +14,14 @@ Invoking "readableTime(3690)" should return "01:01:30" (HH:MM:SS)
 const readableTime = (seconds) => {
   // YOUR CODE HERE...
+  let hours = Math.floor(seconds / 3600).toString().padStart(2, '0'); //calculates the number of full hours in the given seconds.
+  let minutes = Math.floor((seconds % 3600) / 60).toString().padStart(2, '0');// the same as above but divide the remaining seconds by 60 to get the minutes
```
I think the remaining seconds of an hours are used in other operations, so probably you can have a constant to handle that.

erickvh commented on JS-Algorithms/challenges.js:
```diff
@@ -14,6 +14,14 @@ Invoking "readableTime(3690)" should return "01:01:30" (HH:MM:SS)
   // YOUR CODE HERE...
+  let hours = Math.floor(seconds / 3600).toString().padStart(2, '0'); //calculates the number of full hours in the given seconds.
+  let minutes = Math.floor((seconds % 3600) / 60).toString().padStart(2, '0');// the same as above but divide the remaining seconds by 60 to get the minutes
+  let secs = Math.floor(seconds % 3600 % 60).toString().padStart(2, '0');//literally the same as above but obtain the module of the remaining seconds by 60 to get the seconds
```
In order to explain better you should create constants to indicate what `3600` and `60` mean.

erickvh commented on JS-Algorithms/challenges.js:
```diff
@@ -14,6 +14,14 @@ Invoking "readableTime(3690)" should return "01:01:30" (HH:MM:SS)
 
 const readableTime = (seconds) => {
   // YOUR CODE HERE...
+  let hours = Math.floor(seconds / 3600).toString().padStart(2, '0'); //calculates the number of full hours in the given seconds.
```
You can use const since you won't re-assign a different number in the variable

erickvh commented on JS-Algorithms/challenges.js:
```diff
@@ -42,12 +50,19 @@ const COUNTRY_NAMES = ["Germany", "Norway", "Island", "Japan", "Israel"];
+  let secondPart = COUNTRY_NAMES.slice(0, index); //get the second part of the array to pass at the beginning
+  // console.log(firstPart)
+  // console.log(secondPart)
+  return firstPart.concat(secondPart); //just concatenate both parts of the array 
```
It is not necessary to store the part since you're not going to use them in other parts of the code, use array spreading instead

```suggestion
   return [...COUNTRY_NAMES.slice(index), ...COUNTRY_NAMES.slice(0, index)];
```

erickvh commented on JS-Algorithms/challenges.js:
```diff
@@ -42,12 +50,19 @@ const COUNTRY_NAMES = ["Germany", "Norway", "Island", "Japan", "Israel"];
 circularArray(3);
 circularArray(5);
 circularArray(9);
+circularArray(9);
```
Why do we duplicate another execution?

erickvh commented on JS-Algorithms/challenges.js:
```diff
@@ -71,8 +86,18 @@ The last 3 digits for the sum of powers from 1 to 10 is "317"
 
 const ownPower = (number, lastDigits) => {
   // YOUR CODE HERE...
+  let suma = BigInt(0); 
```
Try to don't mix english and spanish when adding variables, btw this can be a const

erickvh commented on JS-Algorithms/challenges.js:
```diff
@@ -96,6 +121,16 @@ Since 10! === 3628800 and you sum 3 + 6 + 2 + 8 + 8 + 0 + 0
+    factorial *= BigInt(i);
+  }
+  let fact = 0;
+  factorial.toString().split('').forEach(digit => fact += parseInt(digit));
```
You can avoid doing the split by just using the the string in a for loop

erickvh commented on JS-Algorithms/challenges.js:
```diff
@@ -96,6 +121,16 @@ Since 10! === 3628800 and you sum 3 + 6 + 2 + 8 + 8 + 0 + 0
+  for (let i = 1; i <= n; i++) {
+    factorial *= BigInt(i);
+  }
+  let fact = 0;
```
Probably you can choose another option for the name, since it is quite similar to the `factorial` one

erickvh commented on JS-Algorithms/challenges.js:
```diff
@@ -119,6 +154,16 @@ Because the 12th index in the Fibonacci sequence is 144, and 144 has three digit
 
 const fibIndex = (n) => {
   // YOUR CODE HERE...
+  let fib = [BigInt(0), BigInt(1)];
```
You can improve readability by improving the naming in this constant.

Although I think you should separate element n and n+1

erickvh replied:
☝🏻 Probably an array could not be the best fit here.

erickvh commented on JS-Algorithms/challenges.js:
```diff
@@ -119,6 +154,16 @@ Because the 12th index in the Fibonacci sequence is 144, and 144 has three digit
+  while (true) {
+    fib[index] = fib[index - 1] + fib[index - 2];
+    if (fib[index].toString().length === n) {
+      // console.log(fib)
```
Remove comments to remove code

erickvh commented on JS-Algorithms/challenges.js:
```diff
@@ -14,6 +14,14 @@ Invoking "readableTime(3690)" should return "01:01:30" (HH:MM:SS)
 
 const readableTime = (seconds) => {
```
In this solution you can also use [Remainder assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Remainder_assignment) in case you're curious how it works

erickvh commented on Front-End/challenge1/script.js:
```diff
@@ -1 +1,15 @@
+    let counter = document.getElementById('counter');
+    let currentDecrease = parseInt(counter.textContent);
+    counter.textContent = currentDecrease - 1;
+    currentDecrease <=0 ? alert('Numeros  negativos negativos'): null;
```
I think the validation must follow a different order in the execution to work correctly.

erickvh commented on Front-End/challenge2/script.js:
```diff
@@ -1,7 +1,115 @@
+        handleOperator('+');
+    } else if (key === '-') {
+        handleOperator('-');
+    } else if (key === '*') {
+        handleOperator('X');
```
I think you can handle some consistency in here 
```suggestion
    } else if (key === '*') {
        handleOperator('*');
```

erickvh replied:
And then you can improve all of the condition by

```else if (['-','+', ...].includes(key))```

Menendez2004 committed:
correctios
constant uses
change come methods and some ways to do the functions
unused clg deleted
correction with the variables in spanish
